Platform/OpenNtfsDxe/Info.c

```diff
@@ -0,0 +1,238 @@
+    if (BaseMftRecord->FileRecord == NULL) {
+      DEBUG ((DEBUG_INFO, "Failed to allocate buffer for FileRecord.\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    Status = ReadMftRecord (BaseMftRecord->File, BaseMftRecord->FileRecord, BaseMftRecord->Inode);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "NTFS: Could not read Mft Record #%d.\n", BaseMftRecord->Inode));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return Status;
+    }
+  }
+
+  Status = InitAttr (&BaseMftRecord->Attr, BaseMftRecord);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not initialize attribute.\n"));
+    FreeFile (BaseMftRecord);
+    FreePool (BaseMftRecord);
+    return Status;
+  }
+
+  Res = (ATTR_HEADER_RES *) FindAttr (&BaseMftRecord->Attr, AT_VOLUME_NAME);
+  if ((Res != NULL) && (Res->NonResFlag == 0) && (Res->InfoLength != 0)) {
+    //
+    // The Volume Name is not terminated with a Unicode NULL.
+    // Its name's length is the size of the attribute as stored in the header.
+    //
+    *Label = AllocateZeroPool (Res->InfoLength + sizeof (CHAR16));
+    if (*Label == NULL) {
+      DEBUG ((DEBUG_INFO, "Failed to allocate buffer for *Label\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    CopyMem (*Label, (UINT8 *) Res + Res->InfoOffset, Res->InfoLength);
+  }
+
+  FreeFile (BaseMftRecord);
+  FreePool (BaseMftRecord);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Returns file info, system info, or the label of a volume
+  depending on information type specified.
+
+  @param  This   Pointer to this instance of file protocol
+  @param  Type   Pointer to information type requested
+  @param  Len    Pointer to size of buffer
+  @param  Data   Pointer to buffer for returned info
+
+  @retval EFI_STATUS  Status of the operation
+**/
+EFI_STATUS
+EFIAPI
+FileGetInfo (
+  IN EFI_FILE_PROTOCOL *This,
+  IN EFI_GUID          *Type,
+  IN OUT UINTN         *Len,
+  OUT VOID             *Data
+  )
+{
+  EFI_STATUS                    Status;
+  EFI_NTFS_FILE                 *File;
+  EFI_FILE_INFO                 *Info;
+  EFI_FILE_SYSTEM_INFO          *FSInfo;
+  EFI_FILE_SYSTEM_VOLUME_LABEL  *VLInfo;
+  EFI_TIME                      Time;
+  CHAR16                        *Label;
+  UINTN                         Length = 0;
+
+  ASSERT (This != NULL);
+  ASSERT (Type != NULL);
+  ASSERT (Len != NULL);
+  ASSERT ((Data != NULL) || ((Data == NULL) && (*Len == 0)));
+
+  File = (EFI_NTFS_FILE *) This;
+
+  if (CompareMem (Type, &gEfiFileInfoGuid, sizeof (*Type)) == 0) {
+    //
+    // Fill in file information
+    //
+    Info = (EFI_FILE_INFO *) Data;
+
+    if (File->BaseName != NULL) {
+      Length = sizeof (EFI_FILE_INFO) + StrSize (File->BaseName);
+    } else if (File->Path != NULL) {
+      Length = sizeof (EFI_FILE_INFO) + StrSize (File->Path);
```

Comment:
``EFI_FILE_INFO`` has that borked 1-sized array for FileName, maybe use ``SIZE_OF_EFI_FILE_INFO``. This should not cause any issues with padding right here, but for some other definitions in EDK II it might.